processing/src/main/java/org/apache/druid/frame/write/cast/ObjectToDoubleColumnValueSelector.java

```diff
@@ -0,0 +1,127 @@
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.druid.frame.write.cast;
+
+import org.apache.druid.common.config.NullHandling;
+import org.apache.druid.math.expr.ExprEval;
+import org.apache.druid.query.monomorphicprocessing.RuntimeShapeInspector;
+import org.apache.druid.segment.ColumnValueSelector;
+import org.apache.druid.segment.RowIdSupplier;
+
+import javax.annotation.Nullable;
+
+public class ObjectToDoubleColumnValueSelector implements ColumnValueSelector<Double>
+{
+  private final ColumnValueSelector<?> objectSelector;
+
+  @Nullable
+  private final RowIdSupplier rowIdSupplier;
+
+  @Nullable
+  private Double currentValue;
+  private long currentRowId = RowIdSupplier.INIT;
+
+  public ObjectToDoubleColumnValueSelector(
+      final ColumnValueSelector<?> objectSelector,
+      @Nullable final RowIdSupplier rowIdSupplier
+  )
+  {
+    this.objectSelector = objectSelector;
+    this.rowIdSupplier = rowIdSupplier;
+  }
+
+  @Override
+  public double getDouble()
+  {
+    final Number n = computeIfNeeded();
+    return n == null ? NullHandling.ZERO_DOUBLE : n.doubleValue();
+  }
+
+  @Override
+  public float getFloat()
+  {
+    final Number n = computeIfNeeded();
+    return n == null ? NullHandling.ZERO_FLOAT : n.floatValue();
+  }
+
+  @Override
+  public long getLong()
+  {
+    final Number n = computeIfNeeded();
+    return n == null ? NullHandling.ZERO_LONG : n.longValue();
+  }
+
+  @Override
+  public boolean isNull()
+  {
+    return computeIfNeeded() == null;
+  }
+
+  @Nullable
+  @Override
+  public Double getObject()
+  {
+    return computeIfNeeded();
+  }
+
+  @Override
+  public Class<Double> classOfObject()
+  {
+    return Double.class;
+  }
+
+  @Override
+  public void inspectRuntimeShape(RuntimeShapeInspector inspector)
+  {
+    inspector.visit("objectSelector", objectSelector);
+    inspector.visit("rowIdSupplier", rowIdSupplier);
+  }
+
+  @Nullable
+  private Double computeIfNeeded()
+  {
+    if (rowIdSupplier == null) {
+      return eval();
+    } else {
+      final long rowId = rowIdSupplier.getRowId();
+
+      if (currentRowId != rowId) {
+        currentValue = eval();
+        currentRowId = rowId;
+      }
+
+      return currentValue;
+    }
+  }
+
+  @Nullable
+  private Double eval()
+  {
+    final Object obj = objectSelector.getObject();
+    if (obj == null) {
+      return null;
+    } else if (obj instanceof Number) {
+      return ((Number) obj).doubleValue();
+    } else {
+      final ExprEval<?> eval = ExprEval.bestEffortOf(obj);
+      return eval.isNumericNull() ? null : eval.asDouble();
```

Comment:
Should we through a parse exception here ?

Going though the exprEval code, it seems that if a string expression eval is returned and its not a valid double, then a 0.0D is returned. 
Shoudn't we throw an exception and force the user to add a cast to is sql statment ?

Reply:
After discussion with @clintropolis , the automatic type conversion is used all over the processing stack and the ingestion stack. Hence, parseException may not be correct.

Reply:
Yeah, this sort of coercion behavior is standard in the query stack when reading from selectors that don't match the desired type. It's how we accomplish schema evolution.